src/cli_utils.cc

```diff
@@ -78,69 +78,143 @@ Status LoadYAMLFile(const string& filename, YAML::Node &node) {
 
     return Status::OK();
 }
 
-Status contigs_alleles_of_yaml(const YAML::Node& yaml,
-                               std::vector<std::pair<std::string,size_t> > &contigs,
-                               discovered_alleles &dsals) {
-    Status s;
-    if (!yaml.IsMap()) {
-        return Status::Invalid("not a map at top level");
+
+// In a YAML document built out of top-level sub-documents, get the next document.
+//
+// Notes:
+// -  Catch any exceptions, and return an empty node.
+static YAML::Node yaml_get_next_document(std::istream &is, bool skip_first_line = false) {
+    try {
+        char buf[200];
+        stringstream ss;
+        if (skip_first_line) {
+            // the document starts with a "---", skip it
+            is.getline(buf, 200);
+        }
+
+        while (is.good() && !is.eof()) {
+            is.getline(buf, 200);
+            size_t buf_len = strlen(buf);
+
+            if (buf_len == 3) {
+                // check if we reached the end of this top level document
+                string marker(buf);
+                if (marker == "---" || marker == "...")
+                    break;
+            }
+            ss.write(buf, buf_len);
+            ss.write("\n", 1);
+        }
+
+        // We have the entire document in memory. Convert to
+        // a YAML node and return.
+        return std::move(YAML::Load(ss.str()));
+    } catch (exception e) {
+        return YAML::Node();
     }
+}
+
 
+Status discovered_alleles_of_yaml_stream(std::istream &is,
+                                         std::vector<std::pair<std::string,size_t> > &contigs,
+                                         discovered_alleles &dsals) {
+    Status s;
     contigs.clear();
     dsals.clear();
 
-    // read contigs
-    auto n_contigs = yaml["contigs"];
-    if (!n_contigs.IsSequence()) {
-        return Status::Invalid("contigs should be a yaml sequence");
-    }
-    for (auto p = n_contigs.begin(); p != n_contigs.end(); ++p) {
-        const std::string name = (*p)["name"].as<std::string>();
-        size_t size = (*p)["size"].as<size_t>();
-        contigs.push_back(make_pair(name, size));
-    }
+    // The first top-level document is the contigs
+    YAML::Node doc = yaml_get_next_document(is, true);
+    S(contigs_of_yaml(doc, contigs));
+
+    // All other documents are discovered-alleles
+    for (doc = yaml_get_next_document(is);
+         !doc.IsNull();
+         doc = yaml_get_next_document(is)) {
+        allele allele(range(-1,-1,-1), "A");
+        discovered_allele_info ainfo;
 
-    // read ranges and alleles
-    auto n_ranges = yaml["alleles"];
-    s = discovered_alleles_of_yaml(n_ranges, contigs, dsals);
-    if (s.bad()) return s;
+        S(one_discovered_allele_of_yaml(doc, contigs, allele, ainfo));
+        dsals[allele] = ainfo;
+    }
+    if (contigs.size() == 0)
+        return Status::Invalid("Empty contigs");
+    if (dsals.size() == 0)
+        return Status::Invalid("empty discovered alleles");
```

Comment:
I think it may be rare but valid to have zero discovered alleles; would be comfortable removing this check if we also have the check for the "..." EOF marker.